torchao/prototype/float8nocompile/kernels/fp8_dynamic_tensorwise.py

```diff
@@ -31,8 +31,86 @@
 }
 
 
+class KernelAlgorithm(Enum):
+    """Enum for FP8 conversion strategy."""
+
+    # use atomic max to compute global amax between blocks
+    ATOMIC_MAX = "atomic_max"
+
+    # reduce shared buffer containing local block amaxes to find global amax
+    REDUCTION = "reduction"
+
+
+kernel_configs = [
+    triton.Config({"BLOCK_SIZE": 128}, num_warps=1),
+    triton.Config({"BLOCK_SIZE": 256}, num_warps=2),
+    triton.Config({"BLOCK_SIZE": 512}, num_warps=4),
+]
+
+
+# --- atomic max version of kernel ---
+@triton.autotune(configs=kernel_configs, key=["input_size"])
+@triton.jit
+def _block_amax_atomic(
+    input_ptr,
+    amax_ptr,
+    num_elements,
+    input_dtype: tl.constexpr,
+    BLOCK_SIZE: tl.constexpr,
+    EPS: tl.constexpr,
+):
+    # compute local amax for each block
+    block_id = tl.program_id(axis=0)
+    block_start = block_id * BLOCK_SIZE
+    block_offs = block_start + tl.arange(0, BLOCK_SIZE)
+    block_mask = block_offs < num_elements
+    vals = tl.load(input_ptr + block_offs, mask=block_mask).to(input_dtype)
+    block_amax = tl.max(tl.abs(vals), axis=0)
+    tl.atomic_max(amax_ptr, block_amax)
+
+
+@triton.autotune(configs=kernel_configs, key=["input_size"])
+@triton.jit
+def _to_fp8_atomic(
+    input_ptr,
+    scale_out_ptr,
+    amax_ptr,
+    out_ptr,
+    num_elements,
+    fp8_dtype_min,
+    fp8_dtype_max,
+    input_dtype: tl.constexpr,
+    output_dtype: tl.constexpr,
+    BLOCK_SIZE: tl.constexpr,
+    EPS: tl.constexpr,
+):
+    # compute scale, must be fp32
+    global_amax = tl.load(amax_ptr)
+    scale = (fp8_dtype_max / tl.clamp(global_amax, min=EPS, max=float("inf"))).to(
+        tl.float32
+    )
+
+    # only one program needs to store the scale
+    block_id = tl.program_id(axis=0)
+    if block_id == 0:
```

Comment:
not sure why this is needed, maybe delete since doing per-program custom logic increases complexity?

if you want to expose the scale to be able to test it numerically, IMO having your atomics kernel output the scale is simpler

Reply:
>not sure why this is needed, maybe delete since doing per-program custom logic increases complexity?

Since all N thread blocks / program ids will calculate the same scale value for the tensor, we only need 1 of them to store the scale output. I figured having all N programs write the same value to the same memory location would be redundant. 

>if you want to expose the scale to be able to test it numerically, IMO having your atomics kernel output the scale is simpler

I need the scale to use in the Float8Tensor constructor, but yeah I agree it would be simpler to have the atomic amax kernel do the scale calculation right there rather than storing the amax then using it to calculate the scale here. I will refactor.

Reply:
> I need the scale to use in the Float8Tensor constructor

that's a good reason, just put that in the comment

> I figured having all N programs write the same value to the same memory location would be redundant.

I think having a triton program have if statements which branch by program ID is a pattern to avoided (even if it works) because it's suprising / isn't really how people commonly write triton kernels.  IMO we can just refactor your atomic kernel code to also write out the scale to get to the same place with simpler code.

Reply:
> > I need the scale to use in the Float8Tensor constructor
> 
> that's a good reason, just put that in the comment

Done

> > I figured having all N programs write the same value to the same memory location would be redundant.
> 
> I think having a triton program have if statements which branch by program ID is a pattern to avoided (even if it works) because it's suprising / isn't really how people commonly write triton kernels. IMO we can just refactor your atomic kernel code to also write out the scale to get to the same place with simpler code.

Ok good to know, thanks - refactored to remove this pattern and simplify the code, let me know what you think.